core/core.js

```diff
@@ -1,411 +1,123 @@
-      u8Array.byteOffset + u8Array.byteLength - headerByteLength,
-      headerByteLength,
-    );
-
-    const requestId = Number(view.getBigUint64(0, true));
-    const status = view.getUint32(8, true);
-    const result = view.getUint32(12, true);
-
-    // Error handling
-    if (status !== 0) {
-      const className = core.decode(u8Array.subarray(0, result));
-      const message = core.decode(u8Array.subarray(result, -headerByteLength))
-        .trim();
-
-      return [requestId, null, handleError(className, message)];
-    }
-
-    if (u8Array.byteLength === headerByteLength) {
-      return [requestId, result, null];
-    }
-
-    // Rest of response buffer is passed as reference or as a copy
-    let respBuffer = null;
-    if (isCopyNeeded) {
-      // Copy part of the response array (if sent through shared array buf)
-      respBuffer = u8Array.slice(0, result);
+  function opAsyncHandler(promiseId, res) {
+    // const [ok, err] = res;
+    const promise = promiseTable.get(promiseId);
+    promiseTable.delete(promiseId);
+    if (!res[1]) {
+      promise.resolve(res[0]);
     } else {
-      // Create view on existing array (if sent through overflow)
-      respBuffer = u8Array.subarray(0, result);
-    }
-
-    return [requestId, respBuffer, null];
-  }
-
-  function binOpAsync(opName, argument = 0, ...zeroCopy) {
-    return opAsync(
-      opName,
-      (requestId) => binOpBuildRequest(requestId, argument, zeroCopy),
-      binOpParseResult,
-    );
-  }
-
-  function binOpSync(opName, argument = 0, ...zeroCopy) {
-    return opSync(
-      opName,
-      () => binOpBuildRequest(0, argument, zeroCopy),
-      binOpParseResult,
-    );
-  }
-
-  ////////////////////////////////////////////////////////////////////////////////////////////
-  ///////////////////////////////////// Json ops handling ////////////////////////////////////
-  ////////////////////////////////////////////////////////////////////////////////////////////
-
-  const jsonRequestHeaderLength = 8;
-
-  function jsonOpBuildRequest(requestId, argument, zeroCopy) {
-    const u8Array = core.encode(
-      "\0".repeat(jsonRequestHeaderLength) + JSON.stringify(argument),
-    );
-    new DataView(u8Array.buffer).setBigUint64(0, BigInt(requestId), true);
-    return [u8Array, ...zeroCopy];
-  }
-
-  function jsonOpParseResult(u8Array, _) {
-    const data = JSON.parse(core.decode(u8Array));
-
-    if ("err" in data) {
-      return [
-        data.requestId,
-        null,
-        handleError(data.err.className, data.err.message),
-      ];
+      promise.reject(processErr(res[1]));
     }
-
-    return [data.requestId, data.ok, null];
   }
 
-  function jsonOpAsync(opName, argument = null, ...zeroCopy) {
-    return opAsync(
-      opName,
-      (requestId) => jsonOpBuildRequest(requestId, argument, zeroCopy),
-      jsonOpParseResult,
-    );
+  function binOpSync(opName, args = null, zeroCopy = null) {
+    return jsonOpSync(opName, args, zeroCopy);
```

Comment:
This is surprising, can you explain why `binOpSync` actually calls `jsonOpSync`?

Reply:
Implementations are now identical, but relevant to keep as semantic groups.

Might clean up in a follow-up PR, also wanted to minimize code changes outside of `core/` for this PR.